lib/model/autocomplete.dart

```diff
@@ -254,22 +265,43 @@ class MentionAutocompleteView extends ChangeNotifier {
     notifyListeners();
   }
 
+  List<User>? _sortedUsers;
+
+  List<User> sortByRelevance({required List<User> users}) {
+    return users;
+  }
+
+  void _sortUsers() {
+    final users = store.users.values.toList();
+    _sortedUsers = sortByRelevance(users: users);
+  }
+
   Future<List<MentionAutocompleteResult>?> _computeResults(MentionAutocompleteQuery query) async {
     final List<MentionAutocompleteResult> results = [];
-    final Iterable<User> users = store.users.values;
 
-    final iterator = users.iterator;
+    if (_sortedUsers == null) {
+      _sortUsers();
+    }
+
+    final sortedUsers = _sortedUsers!;
+    final iterator = sortedUsers.iterator;
     bool isDone = false;
     while (!isDone) {
       // CPU perf: End this task; enqueue a new one for resuming this work
       await Future(() {});
 
+      if (_sortedUsers != sortedUsers) {
+        // The list of users this loop has been working from has become stale.
+        // Abort so _startSearch can retry with the new list.
+        throw ConcurrentModificationError();
+      }
```

Comment:
I think `ConcurrentModificationError` is fine here. Conceptually, what we are doing is iterating through a collection that got modified: `_sortedUsers` can be thought of as a collection which is a view (in [the database sense](https://en.wikipedia.org/wiki/View_(SQL))) on `store.users`. Because `store.users` got modified, effectively `_sortedUsers` did too.

It's true that the way that "users sorted for autocomplete" collection is *implemented* is that there's a particular `List<User>` (so it's a "materialized view" in the database sense), and iterating through the collection iterates through that, and when the underlying "users" collection gets modified we throw away the list and replace it. That `List<User>` didn't get modified. But that list is an implementation detail in the "users sorted for autocomplete" collection that we're conceptually iterating through.

Reply:
> It would be good to exercise all the different things that can cause a retry…hmm. Maybe […]

Yeah, this would be a good change. We're now depending on those `_sortedUsers = null;` lines in order for this to have the right behavior; so we should extend the tests so that those are exercised. We should aim to make it so that if any one of those lines were deleted, there's some test that would fail.

Reply:
@gnprice, @chrisbobbe 
There's this one thing that I just noticed recently about this part and that is when `store.users` is modified, the `_computeResults` is called twice, once when `_startSearch` is called through `refreshStaleUserResults` and secondly when `ConcurrentModificationError` is thrown which causes a retry.

So what I recommend is the following change in `_startSearch`: _(and I think that's what the `TODO backoff?` comment is there for, if I am not wrong!)_

```diff
-   while (true) {
      try {
        newResults = await _computeResults(query);
-       break;
      } on ConcurrentModificationError {
        // Retry
        // TODO backoff?
+       return;
      }
-   }
```

Reply:
Hmm interesting. Is that a problem already in main?

If so, let's understand it there, and fix it in a prep commit. That reduces the amount of complexity we're debugging at a time.

If not, then what is this code doing differently from what's in main that causes this problem?

Reply:
> Hmm interesting. Is that a problem already in main?

Yeah, this call duplication existed from the main. Also, no `ConcurrentModificationError` would be thrown on `RealmUserUpdateEvent`. Both of them are fixed in the same commit.

Reply:
Cool. In the new revision, it looks like this is the relevant commit — is that right?
7354ce7de autocomplete: Terminate the already-running search when there is a new one

That commit would definitely benefit from some more explanation in the commit message about what problem it's solving and how. See our docs on commit descriptions:
https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-description

Reply:
Added the description.